src/compiler/transformer.ts

```diff
@@ -170,143 +203,233 @@ namespace ts {
-                    }
+            // As a performance optimization, use the cached value of the most recent node.
+            // This helps for cases where this function is called repeatedly for the same node.
+            if (lastNodeEmitFlagsNode === node) {
+                return lastNodeEmitFlags;
+            }
 
-                    return options.flags;
+            // Get the emit flags for a node or from one of its original nodes.
+            let flags: NodeEmitFlags;
+            let current = node;
+            while (current) {
+                if (current.transformId === transformId) {
+                    const nodeEmitFlags = current.emitFlags;
+                    if (nodeEmitFlags) {
+                        flags = nodeEmitFlags & ~NodeEmitFlags.HasNodeEmitFlags;
+                        break;
+                    }
                 }
 
-                node = node.original;
+                current = current.original;
             }
 
-            return undefined;
+            // Cache the most recently requested value.
+            lastNodeEmitFlagsNode = node;
+            lastNodeEmitFlags = flags;
+            return flags;
         }
 
         /**
          * Sets flags that control emit behavior of a node.
+         *
+         * @param node The node.
+         * @param emitFlags The NodeEmitFlags for the node.
          */
-        function setNodeEmitFlags<T extends Node>(node: T, flags: NodeEmitFlags) {
-            const options = getEmitOptions(node, /*create*/ true);
-            if (flags & NodeEmitFlags.Merge) {
-                flags = options.flags | (flags & ~NodeEmitFlags.Merge);
+        function setNodeEmitFlags<T extends Node>(node: T, emitFlags: NodeEmitFlags) {
+            // Merge existing flags.
+            if (emitFlags & NodeEmitFlags.Merge) {
+                emitFlags = getNodeEmitFlags(node) | (emitFlags & ~NodeEmitFlags.Merge);
             }
 
-            options.flags = flags;
+            beforeSetAnnotation(node);
+
+            // Cache the most recently requested value.
+            lastNodeEmitFlagsNode = node;
+            lastNodeEmitFlags = emitFlags;
+            node.emitFlags = emitFlags | NodeEmitFlags.HasNodeEmitFlags;
             return node;
         }
 
         /**
          * Gets a custom text range to use when emitting source maps.
+         *
+         * If a node does not have its own custom source map text range, the custom source map
+         * text range of its original pointer is used.
+         *
+         * @param node The node.
          */
         function getSourceMapRange(node: Node) {
+            // As a performance optimization, use the cached value of the most recent node.
+            // This helps for cases where this function is called repeatedly for the same node.
+            if (lastSourceMapRangeNode === node) {
+                return lastSourceMapRange || node;

```

Comment:
`lastSourceMapRange` may be `undefined`. If there is no custom source map range, we return the node itself.